build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java

```diff
@@ -106,11 +106,10 @@ private static void disableTransitiveDependenciesForSourceSet(Project project, S
      */
     public static void configureCompile(Project project) {
         project.getExtensions().getExtraProperties().set("compactProfile", "full");
-
         JavaPluginExtension java = project.getExtensions().getByType(JavaPluginExtension.class);
+        BuildParams.getJavaToolChainSpec().map(spec -> java.toolchain(spec)).getOrNull();
```

Comment:
Having side-effects in a `map()` call seems a bit like bad practice. I think being explicit here and just putting this behind a call to `ifPresent()` seems more idiomatic.

Reply:
I wish gradle would have a more functional api here. Changed.

Reply:
Agreed. Might be a good simple submission. An `ifPresent()` method like exists on `java.util.Optional` that accepts a `Consumer` would be super useful.